### PR TITLE
Composition of `ValueChoice`

### PR DESCRIPTION
**Pull request by ultmaster**
### Description ###

This PR adds support of composition (transformation) of one or several value choices. For example,

```python
nn.ValueChoice([1, 2]) + nn.ValueChoice([3, 4]) + 5
```

Example use cases that will show in docs when #4399 is merged.

For example, search for a channel scale factor.

```python
class Module:
  scale = nn.ValueChoice([1.0, 1.5, 2.0])
  self.conv1 = nn.Conv2d(3, round(scale * 16))
  self.conv2 = nn.Conv2d(round(scale * 16), round(scale * 64))
  self.conv3 = nn.Conv2d(round(scale * 64), round(scale * 256))

  # forward
  return self.conv3(self.conv2(self.conv1(x)))
```

Or when kernel size and padding are coupled, in order to keep the output size constant.

```python
class Module:
  ks = nn.ValueChoice([3, 5, 7])
  self.conv = nn.Conv2d(3, 16, kernel_size=ks, padding=(ks - 1) // 2)

  # forward
  return self.conv(x)
```

Or when several layers are concatenated for a final layer.

```python
class Module:
  self.linear1 = nn.Linear(3, nn.ValueChoice([1, 2, 3], label='a'))
  self.linear2 = nn.Linear(3, nn.ValueChoice([4, 5, 6], label='b'))
  self.final = nn.Linear(nn.ValueChoice([1, 2, 3], label='a') + nn.ValueChoice([4, 5, 6], label='b'), 2)

  # forward
  return self.final(torch.cat([self.linear1(x), self.linear2(x)], 1))
```

### Checklist ###
  - [x] test case
  - [ ] doc: pending #4399 to avoid conflicts.

### How to test ###

See `test_highlevel_apis.py`.

Adding a full example will be the next task.

### CODE REVIEW

**Comment by QuanluZhang** [comment]
@ultmaster  in which cases this composition is useful?

File: nni/retiarii/nn/pytorch/api.py [hunk 1]
```diff
@@ -228,7 +229,418 @@ def __repr__(self):
...
 
 
-class ValueChoice(Translatable, Mutable):
+class ChosenInputs(nn.Module):
+    """
+    A module that chooses from a tensor list and outputs a reduced tensor.
+    The already-chosen version of InputChoice.
+    """
+
+    def __init__(self, chosen: Union[List[int], int], reduction: str):
```
Comment (QuanluZhang) [inline nni/retiarii/nn/pytorch/api.py:238]: docstring to explain `chosen` and `reduction`

File: nni/retiarii/nn/pytorch/api.py [hunk 2]
```diff
@@ -228,7 +229,418 @@ def __repr__(self):
...
+            return sum(tensor_list)
+        if reduction_type == 'mean':
+            return sum(tensor_list) / len(tensor_list)
+        if reduction_type == 'concat':
+            return torch.cat(tensor_list, dim=1)
+        raise ValueError(f'Unrecognized reduction policy: "{reduction_type}"')
+
+
+# the code in ValueChoice can be generated with this codegen
+# this is not done online because because I want to have type-hint supports
```
Comment (QuanluZhang) [inline nni/retiarii/nn/pytorch/api.py:263]: because because -> because

File: nni/retiarii/nn/pytorch/api.py [hunk 3]
```diff
@@ -228,7 +229,418 @@ def __repr__(self):
...
+        for arg in self.arguments:
+            if isinstance(arg, ValueChoiceX):
+                # recursive evaluation
+                eval_args.append(arg._evaluate(values, dry_run))
+                # the recursion will stop when it hits a leaf node (value choice)
+                # the implementation is in `ValueChoice`
+            else:
+                # constant value
+                eval_args.append(arg)
+        return self.function(*eval_args)
```
Comment (QuanluZhang) [inline nni/retiarii/nn/pytorch/api.py:386]: where is `self.function` defined?

**Comment by ultmaster** [comment]
> @ultmaster in which cases this composition is useful?

For example, search for a channel scale factor.

```python
class Module:
  scale = nn.ValueChoice([1.0, 1.5, 2.0])
  self.conv1 = nn.Conv2d(3, round(scale * 16))
  self.conv2 = nn.Conv2d(round(scale * 16), round(scale * 64))
  self.conv3 = nn.Conv2d(round(scale * 64), round(scale * 256))

  # forward
  return self.conv3(self.conv2(self.conv1(x)))
```

Or when kernel size and padding are coupled, in order to keep the output size constant.

```python
class Module:
  ks = nn.ValueChoice([3, 5, 7])
  self.conv = nn.Conv2d(3, 16, kernel_size=ks, padding=(ks - 1) // 2)

  # forward
  return self.conv(x)
```

Or when several layers are concatenated for a final layer.

```python
class Module:
  self.linear1 = nn.Linear(3, nn.ValueChoice([1, 2, 3], label='a'))
  self.linear2 = nn.Linear(3, nn.ValueChoice([4, 5, 6], label='b'))
  self.final = nn.Linear(nn.ValueChoice([1, 2, 3], label='a') + nn.ValueChoice([4, 5, 6], label='b'), 2)

  # forward
  return self.final(torch.cat([self.linear1(x), self.linear2(x)], 1))

File: nni/retiarii/nn/pytorch/mutator.py [hunk 4]
```diff
@@ -63,30 +63,64 @@ def mutate(self, model):
...
+        super().__init__(label=label)
+        self.candidates = candidates
+
+    def mutate(self, model: Model) -> Model:
+        # leave a record here
+        # real mutations will be done in ParameterChoiceMutator
+        self.choice(self.candidates)
+
+
 class ParameterChoiceMutator(Mutator):
```
Comment (QuanluZhang) [inline nni/retiarii/nn/pytorch/mutator.py:94]: better to give more explanation for this class.

Comment (QuanluZhang) [reply nni/retiarii/nn/pytorch/mutator.py:94]: for example, to claim that this is to deal with the ValueChoice that is used as class (basic unit) or function's input argument

File: nni/retiarii/nn/pytorch/mutator.py [hunk 5]
```diff
@@ -143,18 +177,29 @@ def process_inline_mutation(model: Model) -> Optional[List[Mutator]]:
...
+
+    leaf_value_choices: Dict[str, List[Any]] = {}
+    for _, __, choice in pc_nodes:
+        for inner_choice in choice.inner_choices():
+            if inner_choice.label not in leaf_value_choices:
+                leaf_value_choices[inner_choice.label] = inner_choice.candidates
+            else:
+                assert leaf_value_choices[inner_choice.label] == inner_choice.candidates, \
+                    'Value choice with the same label must have the same candidates, but found ' \
+                    f'{leaf_value_choices[inner_choice.label]} vs. {inner_choice.candidates}'
```
Comment (QuanluZhang) [inline nni/retiarii/nn/pytorch/mutator.py:195]: it is hard to understand this part, especially `leaf_value_choices`

File: nni/retiarii/nn/pytorch/mutator.py [hunk 6]
```diff
@@ -234,9 +283,10 @@ def extract_mutation_from_pt_module(pytorch_model: nn.Module) -> Tuple[Model, Op
         # tricky case: value choice that serves as parameters are stored in traced arguments
         if is_basic_unit(module):
             for key, value in module.trace_kwargs.items():
-                if isinstance(value, ValueChoice):
-                    node = graph.add_node(name + '.init.' + key, 'ValueChoice', {'candidates': value.candidates})
-                    node.label = value.label
+                if isinstance(value, ValueChoiceX):
+                    for i, choice in enumerate(value.inner_choices()):
+                        node = graph.add_node(f'{name}.init.{key}.{i}', 'ValueChoice', {'candidates': choice.candidates})
+                        node.label = choice.label
```
Comment (QuanluZhang) [inline nni/retiarii/nn/pytorch/mutator.py:289]: `extract_mutation_from_pt_module` is for python engine, right? the returned model of this function is not a connected graph, right?

Comment (ultmaster) [reply nni/retiarii/nn/pytorch/mutator.py:289]: Yes. The nodes only exist to remind the sampler that they need to be sampled.

**Review by QuanluZhang — APPROVED** [review]

**Review by liuzhe-lz — APPROVED** [review]
